File: Tim/ChainStory.java
package Tim;

import Tim.Commands.ICommandHandler;
import Tim.Data.ChannelInfo;
import Tim.Data.CommandData;
import org.pircbotx.hooks.types.GenericMessageEvent;
import org.pircbotx.hooks.types.GenericUserEvent;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.DecimalFormat;
import java.util.Calendar;
import java.util.logging.Level;
import java.util.logging.Logger;

class ChainStory implements ICommandHandler {
	private final long timeout = 3000;

	public boolean handleCommand(CommandData commandData) {
		ChannelInfo cdata = Tim.db.channel_data.get(commandData.getChannelEvent()
			.getChannel()
			.getName()
			.toLowerCase());
		Calendar cal = Calendar.getInstance();

		boolean isNovember = (Calendar.NOVEMBER == cal.get(Calendar.MONTH));

		switch (commandData.command) {
			case "chainlast":
				if (cdata.commands_enabled.get("chainstory")) {
					showLast(commandData.getUserEvent());
				} else {
					commandData.event.respond("I'm sorry. I don't do that here.");
				}

				return true;
			case "chainnew":
				if (cdata.commands_enabled.get("chainstory")) {
					if (!isNovember) {
						commandData.getMessageEvent()
							.respond("Sorry, that command won't work outside November!");
						return true;
					}
					addNew(commandData);
				} else {
					commandData.getMessageEvent()
						.respond("I'm sorry. I don't do that here.");
				}

				return true;
			case "chaininfo":
				if (cdata.commands_enabled.get("chainstory")) {
					info(commandData.getMessageEvent());
				} else {
					commandData.getMessageEvent()
						.respond("I'm sorry. I don't do that here.");
				}

				return true;
			case "chaincount":
				if (cdata.commands_enabled.get("chainstory")) {
					count(commandData.getMessageEvent());
				} else {
					commandData.getMessageEvent()
						.respond("I'm sorry. I don't do that here.");
				}

				return true;
		}

		return false;
	}

	private void showLast(GenericUserEvent event) {
		if (event.getUser() == null) {
			return;
		}

		Connection con;
		try {
			con = Tim.db.pool.getConnection(timeout);

			PreparedStatement s = con.prepareStatement("SELECT * FROM `story` ORDER BY id DESC LIMIT 3");
			s.executeQuery();

			event.respond("I sent you the last three paragraphs in a private message... They're too awesome for everyone to see!");

			ResultSet rs = s.getResultSet();
			rs.setFetchDirection(ResultSet.FETCH_REVERSE);
			rs.last();
			do {
				event.getUser()
					.send()
					.message(rs.getString("string"));
			} while (rs.previous());

			con.close();
		} catch (SQLException ex) {
			Tim.printStackTrace(ex);
		}
	}

	private void addNew(CommandData commandData) {
		if (commandData.getUserEvent()
			.getUser() == null) {
			return;
		}

		DecimalFormat formatter = new DecimalFormat("#,###");
		Connection con;

		String message = commandData.argString.replaceAll("^<?(.*)>?$", "$1");

		if ("".equals(message)) {
			commandData.getChannelEvent()
				.getChannel()
				.send()
				.action("blinks, and looks confused. \"But there's nothing there. That won't help my wordcount!\"");
		} else {
			try {
				con = Tim.db.pool.getConnection(timeout);

				storeLine(message, commandData.getMessageEvent()
					.getUser()
					.getNick());
				commandData.getChannelEvent()
					.getChannel()
					.send()
					.action("quickly copies down what " + commandData.getUserEvent()
						.getUser()
						.getNick() + " said. \"Thanks!\"");

				PreparedStatement s = con.prepareStatement(
					"SELECT SUM( LENGTH( STRING ) - LENGTH( REPLACE( STRING ,  ' ',  '' ) ) +1 ) AS word_count FROM story");
				s.executeQuery();
				ResultSet rs = s.getResultSet();
				while (rs.next()) {
					commandData.getMessageEvent()
						.respond("My novel is now " + formatter.format(Integer.parseInt(rs.getString("word_count"))) + " words long!");
				}

				con.close();
			} catch (SQLException ex) {
				Tim.printStackTrace(ex);
			}
		}
	}

	String getLastLines() {
		Connection con;
		ResultSet rs;
		PreparedStatement s;
		StringBuilder response = new StringBuilder();

		try {
			con = Tim.db.pool.getConnection(timeout);

			s = con.prepareStatement("SELECT * FROM `story` ORDER BY id DESC LIMIT 3");
			s.executeQuery();

			rs = s.getResultSet();
			rs.setFetchDirection(ResultSet.FETCH_REVERSE);
			rs.last();
			do {
				response.append(rs.getString("string"))
					.append(" ");
			} while (rs.previous());

			con.close();
		} catch (SQLException ex) {
			Tim.printStackTrace(ex);
		}

		return response.toString();
	}

	private void info(GenericMessageEvent event) {
		if (event.getUser() == null) {
			return;
		}

		DecimalFormat formatter = new DecimalFormat("#,###");
		Connection con;
		int word_count = 0, author_count = 0;
		ResultSet rs;
		PreparedStatement s;

		try {
			con = Tim.db.pool.getConnection(timeout);

			s = con.prepareStatement("SELECT SUM( LENGTH( STRING ) - LENGTH( REPLACE( STRING ,  ' ',  '' ) ) +1 ) AS word_count FROM story");
			s.executeQuery();
			rs = s.getResultSet();
			while (rs.next()) {
				word_count = Integer.parseInt(rs.getString("word_count"));
			}

			s = con.prepareStatement("SELECT COUNT(DISTINCT author) AS author_count FROM story");
			s.executeQuery();
			rs = s.getResultSet();
			while (rs.next()) {
				author_count = Integer.parseInt(rs.getString("author_count"));
			}

			event.respond("My novel is currently " + formatter.format(word_count) + " words long, with paragraphs written by " + formatter.format(author_count)
				+ " different people, and I sent you the last three paragraphs are in a private message... They're too awesome for everyone to "
				+ "see!");

			s = con.prepareStatement("SELECT * FROM `story` ORDER BY id DESC LIMIT 3");
			s.executeQuery();

			rs = s.getResultSet();
			rs.setFetchDirection(ResultSet.FETCH_REVERSE);
			rs.last();
			do {
				event.getUser()
					.send()
					.message(rs.getString("string"));
			} while (rs.previous());

			event.respond("You can read an excerpt in my profile here: http://www.nanowrimo.org/en/participants/timmybot");

			con.close();
		} catch (SQLException ex) {
			Tim.printStackTrace(ex);
		}
	}

	private void count(GenericMessageEvent event) {
		DecimalFormat formatter = new DecimalFormat("#,###");
		Connection con;
		int word_count = 0, author_count = 0;
		ResultSet rs;
		PreparedStatement s;

		try {
			con = Tim.db.pool.getConnection(timeout);

			s = con.prepareStatement("SELECT SUM( LENGTH( STRING ) - LENGTH( REPLACE( STRING ,  ' ',  '' ) ) +1 ) AS word_count FROM story");
			s.executeQuery();
			rs = s.getResultSet();
			while (rs.next()) {
				word_count = Integer.parseInt(rs.getString("word_count"));
			}

			s = con.prepareStatement("SELECT COUNT(DISTINCT author) AS author_count FROM story");
			s.executeQuery();
			rs = s.getResultSet();
			while (rs.next()) {
				author_count = Integer.parseInt(rs.getString("author_count"));
			}

			event.respond("My novel is currently " + formatter.format(word_count) + " words long, with paragraphs written by " + formatter.format(author_count)
				+ " different people.");

			con.close();
		} catch (SQLException ex) {
			Tim.printStackTrace(ex);
		}
	}

	void storeLine(String line, String author) {
		Connection con;
		try {
			con = Tim.db.pool.getConnection(timeout);

			PreparedStatement s = con.prepareStatement("INSERT INTO story SET string = ?, author = ?, created = NOW()");
			s.setString(1, line);
			s.setString(2, author);
			s.executeUpdate();

			con.close();
		} catch (SQLException ex) {
			Tim.printStackTrace(ex);
		}
	}

	void helpSection(GenericUserEvent event) {
		String[] strs = {
			"Timmy's chain story is a ridiculous attempt to create a horrific 50,000 word 'novel' out of",
			"contributions from people in chat and autogenerated text. It will be released in December",
			"as a book you can purchase on Amazon, and all contributors will be given credit in the book.",
			"Chain Story Commands:",
			"    !chaininfo - General info about the current status of my novel.",
			"    !chainlast - The last paragraph of my novel, so you have something to base the next one one.",
			"    !chainnew <paragraph> - Provide the next paragraph of my great cyberspace novel!", "    !chaincount - Just the word count and author stats.",
		};

		for (String str : strs) {
			event.getUser()
				.send()
				.message(str);
		}
	}

	int wordcount() {
		Connection con;
		ResultSet rs;
		PreparedStatement s;
		int word_count = 0;

		try {
			con = Tim.db.pool.getConnection(timeout);

			s = con.prepareStatement("SELECT SUM( LENGTH( STRING ) - LENGTH( REPLACE( STRING ,  ' ',  '' ) ) +1 ) AS word_count FROM story");
			s.executeQuery();
			rs = s.getResultSet();
			while (rs.next()) {
				String count = rs.getString("word_count");
				if (count != null && !count.equals("")) {
					word_count = Integer.parseInt(count);
				}
			}

			con.close();
		} catch (SQLException ex) {
			Logger.getLogger(Tim.class.getName())
				.log(Level.SEVERE, null, ex);
		}

		return word_count;
	}
}

File: timmy/command_processors/writing/chainstory.py
import re
from datetime import datetime

from timmy.command_processors.base_command import BaseCommand
from timmy.data.command_data import CommandData


class ChainStoryCommands(BaseCommand):
    user_commands = ["chainstory", "chainlast", "chainnew", "chaininfo", "chaincount"]
    sub_commands = ["last", "new", "info", "count"]

    interaction_checks = False

    help_topics = [('user', 'chain story', '!chain info', 'General info about the current status of my novel.'),
                   ('user', 'chain story', '!chain last', 'The last paragraph of my novel, so you have something to '
                                                          'base the next bit on.'),
                   ('user', 'chain story', '!chain new <paragraph>', 'Provide the next paragraph of my great novel!'),
                   ('user', 'chain story', '!chain count', 'Get the current wordcount and contributor count for my '
                                                           'novel.')]

    command_flag_map = {
        'chainstory': 'chainstory',
        'chainlast':  'chainstory',
        'chainnew':   'chainstory',
        'chaininfo':  'chainstory',
        'chaincount': 'chainstory'
    }

    def process(self, command_data: CommandData) -> None:
        if self._execution_checks(command_data):
            if command_data.command == 'chainlast':
                self.respond_to_user(command_data, "The syntax for chain-related commands has changed. Try '!chainstory"
                                                   " last' in the future.")
                command_data.command = 'chainstory'
                command_data.args.insert(0, 'last')
                command_data.arg_count += 1
                command_data.arg_string = 'last ' + command_data.arg_string
            elif command_data.command == 'chainnew':
                self.respond_to_user(command_data, "The syntax for chain-related commands has changed. Try '!chainstory"
                                                   " new' in the future.")
                command_data.command = 'chainstory'
                command_data.args.insert(0, 'new')
                command_data.arg_count += 1
                command_data.arg_string = 'new ' + command_data.arg_string
            elif command_data.command == 'chaininfo':
                self.respond_to_user(command_data, "The syntax for chain-related commands has changed. Try '!chainstory"
                                                   " info' in the future.")
                command_data.command = 'chainstory'
                command_data.args.insert(0, 'info')
                command_data.arg_count += 1
                command_data.arg_string = 'info ' + command_data.arg_string
            elif command_data.command == 'chaincount':
                self.respond_to_user(command_data, "The syntax for chain-related commands has changed. Try '!chainstory"
                                                   " count' in the future.")
                command_data.command = 'chainstory'
                command_data.args.insert(0, 'count')
                command_data.arg_count += 1
                command_data.arg_string = 'count ' + command_data.arg_string

            self.handle_subcommand(command_data)

    def _last_handler(self, command_data: CommandData) -> None:
        from timmy.db_access import chainstory_db
        from timmy.core import bot_instance

        last_lines = chainstory_db.get_last_lines()

        for line in last_lines:
            bot_instance.connection.privmsg(command_data.issuer, line[:480])

        self.respond_to_user(command_data, "I sent you the last three paragraphs in a private message... They're too "
                                           "awesome for everyone to see!")
        return

    def _new_handler(self, command_data: CommandData) -> None:
        month = datetime.now().month
        if month != 11:
            self.respond_to_user(command_data, "Sorry, that command won't work outside November!")
            return

        message = " ".join(command_data.args[1:])
        message = re.sub(r'^<?(.*?)>?$', r'\1', message)

        if message == "":
            self.respond_to_user(command_data, "blinks, and looks confused. \"But there's nothing there. That won't "
                                               "help my wordcount!\"")
            return

        from timmy.db_access import chainstory_db
        chainstory_db.add_line(message, command_data.issuer)

        word_count = chainstory_db.word_count()

        self.send_action(command_data, f"quickly jots down what {command_data.issuer} said. \"Thanks! My novel is now "
                                       f"{word_count:,} words long!\"")

    def _info_handler(self, command_data: CommandData) -> None:
        from timmy.db_access import chainstory_db
        from timmy.core import bot_instance

        last_lines = chainstory_db.get_last_lines()
        word_count = chainstory_db.word_count()
        author_count = chainstory_db.author_count()

        for line in last_lines:
            bot_instance.connection.privmsg(command_data.issuer, line)

        self.respond_to_user(command_data, f"My novel is currently {word_count:,} words long, with paragraphs written "
                                           f"by {author_count:,} different people, and I sent you the last three "
                                           f"paragraphs in a private message... They're too awesome for everyone to "
                                           f"see!")

        self.respond_to_user(command_data, "You can read an excerpt in my profile here: "
                                           "https://nanowrimo.org/participants/timmybot")
        return

    def _count_handler(self, command_data: CommandData) -> None:
        from timmy.db_access import chainstory_db
        word_count = chainstory_db.word_count()
        author_count = chainstory_db.author_count()

        self.respond_to_user(command_data, f"My novel is currently {word_count:,} words long, with paragraphs written "
                                           f"by {author_count:,} different people.")
        return
